File: frontend/src/features/classroom/submissionLimits.ts
/**
 * The limits a homework submission is held to, and the one place the upload panel reads them.
 *
 * The server owns all of them: `backend/classes/submission_limits.py` and the
 * `CLASSROOM_SUBMISSION_*` settings re-check every one on POST, and a batch that gets past this
 * file is still refused there. So the assignment payload carries them (`submission_limits`) and
 * this module only reshapes what it is told — ops can lower a limit with an environment variable
 * and the panel moves with it. They are checked here as well as there so a student with a 60 MB
 * scan learns it while picking rather than after waiting out an upload that was never going to
 * land, and they live in this module alone so the panel and its tests cannot drift apart.
 */

/** What the server may not accept, as it words it in the assignment payload. */
export interface SubmissionLimitsPayload {
  max_files_per_submission?: number | null;
  max_file_bytes?: number | null;
  max_batch_bytes?: number | null;
  allowed_extensions?: string[] | null;
}

export interface SubmissionLimits {
  /** Files one submission may hold in total — what is already attached plus what is arriving. */
  maxFiles: number;
  /** Largest single file. */
  maxFileBytes: number;
  /** Largest total for the *new* files in one request; what is already attached does not count. */
  maxBatchBytes: number;
  /** Lower-case, leading dot — ".pdf", ".jpg". A file whose name has no dot is not judged here. */
  allowedExtensions: readonly string[];
}

/**
 * Used only until the assignment payload answers — an older backend, or the first paint of a
 * cached page. Each is the default the deployed settings ship with, except the batch figure:
 * Nginx caps an `/api/` body at 60 MB, below Django's 100 MB, and a guess that is too generous
 * is the one that costs a student their upload.
 */
export const FALLBACK_SUBMISSION_LIMITS: SubmissionLimits = {
  maxFiles: 50,
  maxFileBytes: 50 * 1024 * 1024,
  maxBatchBytes: 60 * 1024 * 1024,
  allowedExtensions: [
    ".doc", ".docx", ".gif", ".jpeg", ".jpg", ".pdf", ".png",
    ".ppt", ".pptx", ".txt", ".webp", ".xls", ".xlsx",
  ],
};

function positive(n: unknown, fallback: number): number {
  return typeof n === "number" && Number.isFinite(n) && n > 0 ? n : fallback;
}

/** ".PDF" and "pdf" both mean the same thing to the server, which lower-cases and expects a dot. */
function normalizeExtension(raw: string): string {
  const t = raw.trim().toLowerCase();
  if (!t) return "";
  return t.startsWith(".") ? t : `.${t}`;
}

/** The served limits, or the fallbacks above for whichever the payload does not carry. */
export function resolveSubmissionLimits(served?: SubmissionLimitsPayload | null): SubmissionLimits {
  const extensions = (served?.allowed_extensions ?? [])
    .filter((x): x is string => typeof x === "string")
    .map(normalizeExtension)
    .filter(Boolean);
  return {
    maxFiles: positive(served?.max_files_per_submission, FALLBACK_SUBMISSION_LIMITS.maxFiles),
    maxFileBytes: positive(served?.max_file_bytes, FALLBACK_SUBMISSION_LIMITS.maxFileBytes),
    maxBatchBytes: positive(served?.max_batch_bytes, FALLBACK_SUBMISSION_LIMITS.maxBatchBytes),
    // An empty list would mean "nothing can be turned in", which no deployment intends and which
    // would refuse every file a student picked.
    allowedExtensions: extensions.length > 0 ? extensions : FALLBACK_SUBMISSION_LIMITS.allowedExtensions,
  };
}

/**
 * What makes two picks the same file.
 *
 * A `File` carries no id, and two `File` objects for one file on disk are never `===` — the
 * picker mints a fresh object each time it is opened. Name, size and last-modified time is all a
 * browser can see of a file, and it is enough to tell "picked page1.jpg twice" apart from "picked
 * two different scans that happen to share a folder". Two genuinely distinct files that agree on
 * all three read as one here; that costs a student one re-pick, where the alternative costs them
 * a silent duplicate in front of their teacher.
 */
export function fileIdentity(f: File): string {
  return `${f.name}:${f.size}:${f.lastModified}`;
}

/** Bytes as a student reads them: "12.4 MB", "680 KB", "0 B". */
export function sizeLabel(bytes: number): string {
  if (bytes < 1024) return `${bytes} B`;
  const kb = bytes / 1024;
  if (kb < 1024) return `${Math.round(kb)} KB`;
  const mb = kb / 1024;
  return `${mb >= 10 ? Math.round(mb) : mb.toFixed(1)} MB`;
}

/** ".heic" from "IMG_0421.heic", "" from a name with no dot. Mirrors the server's splitext. */
export function fileExtension(name: string): string {
  const base = name.slice(name.lastIndexOf("/") + 1);
  const dot = base.lastIndexOf(".");
  return dot > -1 ? base.slice(dot).toLowerCase() : "";
}

/** For the picker's `accept`: ".pdf,.png,.jpg". A hint to the file dialog, never a check. */
export function acceptAttribute(limits: SubmissionLimits): string {
  return limits.allowedExtensions.join(",");
}

/** The same list as a student reads it: "PDF, PNG, JPG". */
export function fileTypeList(limits: SubmissionLimits): string {
  return limits.allowedExtensions.map((e) => e.replace(/^\./, "").toUpperCase()).join(", ");
}

export interface BatchCheck {
  /** Queued files of a kind the server will not take, in the order they were picked. */
  disallowed: File[];
  /** Queued files over the per-file limit, in the order they were picked. */
  tooBig: File[];
  /** Total bytes of the queued files — what this request would carry. */
  totalBytes: number;
  /** Files the submission would hold once these are added, attached ones included. */
  totalFiles: number;
  /** One sentence for the student, or null when the batch is ready to go. */
  problem: string | null;
}

/** "One more file is over too." / " 3 more files are over too." — never a bare count of 1. */
function andTheOthers(count: number, one: string, many: (n: number) => string): string {
  if (count === 2) return ` ${one}`;
  if (count > 2) return ` ${many(count - 1)}`;
  return "";
}

/**
 * Check a queue against every limit before anything is sent.
 *
 * The order follows the server's own: it validates each file's type first, then its size, and
 * refuses the whole request on the first bad one — so a student who would lose ten files to a
 * single `.heic` hears about the `.heic`. A problem that names a file is worth more than "the
 * total is too big", which is the same fact stated without a culprit.
 */
export function checkSubmissionBatch(
  queued: File[],
  attachedCount: number,
  limits: SubmissionLimits,
): BatchCheck {
  const allowed = new Set(limits.allowedExtensions);
  // A file with no extension at all passes here because it passes on the server: its check
  // reads `if ext_lower and ext_lower not in allowed`.
  const disallowed = queued.filter((f) => {
    const ext = fileExtension(f.name);
    return ext !== "" && !allowed.has(ext);
  });
  const tooBig = queued.filter((f) => f.size > limits.maxFileBytes);
  const totalBytes = queued.reduce((sum, f) => sum + f.size, 0);
  const totalFiles = attachedCount + queued.length;

  let problem: string | null = null;
  if (disallowed.length > 0) {
    const first = disallowed[0];
    problem =
      `“${first.name}” isn’t a kind of file this takes. These are: ${fileTypeList(limits)}. ` +
      `Save it as one of them and choose it again.` +
      andTheOthers(
        disallowed.length,
        "One more file needs the same.",
        (n) => `${n} more files need the same.`,
      );
  } else if (tooBig.length > 0) {
    const first = tooBig[0];
    problem =
      `“${first.name}” is ${sizeLabel(first.size)}. Each file can be up to ` +
      `${sizeLabel(limits.maxFileBytes)} — save it a little smaller, or split it into a few files.` +
      andTheOthers(
        tooBig.length,
        "One more file is over too.",
        (n) => `${n} more files are over too.`,
      );
  } else if (totalFiles > limits.maxFiles) {
    problem =
      `That would make ${totalFiles} files on this submission. Up to ` +
      `${limits.maxFiles} fit — take a few out and send the rest.`;
  } else if (totalBytes > limits.maxBatchBytes) {
    problem =
      `These come to ${sizeLabel(totalBytes)} together. One upload carries ` +
      `${sizeLabel(limits.maxBatchBytes)} — send some now, then add the rest the same way.`;
  }

  return { disallowed, tooBig, totalBytes, totalFiles, problem };
}

File: ios/MasterSATKit/Sources/MasterSATKit/Homework/SubmissionLimits.swift
import Foundation

/// What one homework hand-in may carry.
///
/// The server refuses anything over these, but only after the upload: a student on mobile
/// data would send 70 MB of photos to be told "File too large" at the end. Checked on the
/// phone first, so the refusal comes before the wait — and in words that say what to do.
///
/// Newer servers send their own numbers as `submission_limits` on the assignment; until then
/// `standard` is the server's own default (`classes/submission_limits.py`): 50 files, 50 MB a
/// file, and 60 MB in one request, because nginx (`client_max_body_size 60M`) refuses a
/// bigger body with an HTML 413 before Django ever sees it.
public struct SubmissionLimits: Decodable, Sendable, Equatable {
    public let maxFilesPerSubmission: Int
    public let maxFileBytes: Int
    public let maxBatchBytes: Int
    /// Lower-case, with the dot: ".pdf".
    public let allowedExtensions: [String]

    public static let standard = SubmissionLimits(
        maxFilesPerSubmission: 50,
        maxFileBytes: 50 * 1024 * 1024,
        maxBatchBytes: 60 * 1024 * 1024,
        allowedExtensions: [".doc", ".docx", ".gif", ".jpeg", ".jpg", ".pdf", ".png",
                            ".ppt", ".pptx", ".txt", ".webp", ".xls", ".xlsx"]
    )

    public init(maxFilesPerSubmission: Int, maxFileBytes: Int, maxBatchBytes: Int, allowedExtensions: [String]) {
        self.maxFilesPerSubmission = maxFilesPerSubmission
        self.maxFileBytes = maxFileBytes
        self.maxBatchBytes = maxBatchBytes
        self.allowedExtensions = allowedExtensions.map(Self.normalise).filter { $0.count > 1 }
    }

    private enum CodingKeys: String, CodingKey {
        case maxFilesPerSubmission = "max_files_per_submission"
        case maxFileBytes = "max_file_bytes"
        case maxBatchBytes = "max_batch_bytes"
        case allowedExtensions = "allowed_extensions"
    }

    /// Any field the server leaves out, or sends as nonsense, keeps the standard value — a
    /// limit of zero would refuse every file.
    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        func positive(_ key: CodingKeys, _ fallback: Int) -> Int {
            let value = (try? c.decodeIfPresent(Int.self, forKey: key)) ?? nil
            return (value ?? 0) > 0 ? value! : fallback
        }
        let standard = Self.standard
        let extensions = ((try? c.decodeIfPresent([String].self, forKey: .allowedExtensions)) ?? nil) ?? []
        self.init(
            maxFilesPerSubmission: positive(.maxFilesPerSubmission, standard.maxFilesPerSubmission),
            maxFileBytes: positive(.maxFileBytes, standard.maxFileBytes),
            maxBatchBytes: positive(.maxBatchBytes, standard.maxBatchBytes),
            allowedExtensions: extensions.isEmpty ? standard.allowedExtensions : extensions
        )
    }

    private static func normalise(_ ext: String) -> String {
        let trimmed = ext.trimmingCharacters(in: .whitespaces).lowercased()
        return trimmed.hasPrefix(".") ? trimmed : "." + trimmed
    }

    // MARK: - Checking

    /// Something that would be refused.
    public enum Problem: Equatable, Sendable {
        case typeNotAccepted(fileName: String)
        case fileTooLarge(fileName: String, bytes: Int)
        case tooManyFiles(total: Int)
        case uploadTooLarge(bytes: Int)
    }

    /// One file, as it is picked: is it a type the server takes, and small enough?
    public func problem(fileName: String, bytes: Int) -> Problem? {
        let ext = Self.extensionOf(fileName)
        if ext.isEmpty || !allowedExtensions.contains(ext) { return .typeNotAccepted(fileName: fileName) }
        if bytes > maxFileBytes { return .fileTooLarge(fileName: fileName, bytes: bytes) }
        return nil
    }

    /// The whole upload, just before it is sent: every file again, the count the submission
    /// would reach, and the size of the one request.
    public func problem(files: [(name: String, bytes: Int)], alreadyHandedIn: Int) -> Problem? {
        for file in files {
            if let problem = problem(fileName: file.name, bytes: file.bytes) { return problem }
        }
        let total = alreadyHandedIn + files.count
        if total > maxFilesPerSubmission { return .tooManyFiles(total: total) }
        let bytes = files.reduce(0) { $0 + $1.bytes }
        if bytes > maxBatchBytes { return .uploadTooLarge(bytes: bytes) }
        return nil
    }

    /// What to tell the student — what happened, and what to do instead.
    public func message(for problem: Problem) -> String {
        switch problem {
        case .typeNotAccepted(let name):
            return "“\(name)” can't be handed in as it is. Files that work: \(allowedExtensions.joined(separator: ", "))."
        case .fileTooLarge(let name, let bytes):
            return "“\(name)” is \(Self.megabytes(bytes)). Each file can be up to \(Self.megabytes(maxFileBytes)) — a smaller photo or a PDF will go through."
        case .tooManyFiles(let total):
            let over = total - maxFilesPerSubmission
            return "A hand-in holds up to \(maxFilesPerSubmission) files, and this would make \(total). Remove \(over) and send again."
        case .uploadTooLarge(let bytes):
            return "These files come to \(Self.megabytes(bytes)), and one upload can carry \(Self.megabytes(maxBatchBytes)). Send some now and the rest after."
        }
    }

    /// ".pdf" from "Essay.PDF"; empty when there is no extension.
    public static func extensionOf(_ fileName: String) -> String {
        let base = fileName.split(separator: "/").last.map(String.init) ?? fileName
        guard let dot = base.lastIndex(of: "."), dot != base.index(before: base.endIndex) else { return "" }
        return String(base[dot...]).lowercased()
    }

    /// "72 MB", "2.4 MB", "680 KB" — the web's `formatBytes`.
    public static func megabytes(_ bytes: Int) -> String {
        if bytes < 1024 { return "\(bytes) B" }
        let kb = Double(bytes) / 1024
        if kb < 1024 { return "\(Int(kb.rounded())) KB" }
        let mb = kb / 1024
        return mb >= 10 ? "\(Int(mb.rounded())) MB" : String(format: "%.1f MB", mb)
    }
}
